src/machine_charm.py

```diff
@@ -93,7 +93,7 @@ def on_remove(self, _) -> None:
 
     def is_ready(self):
         """Checks if the charm is ready for configuration."""
-        return self._is_installed()
+        return self._is_installed() and self.principal_unit
```

Comment:
a bit unclear why there being a principal unit means that the charm is ready. 
Could you add a comment?

Reply:
I will get to it later, but the gist of it is that the is_ready method is used to detect when to actually start writing the configuration files. We need to know the name of the principal unit because it is used to generate the configuration file. I suppose we could also just adapt the config file generators, but at present there are a number of interconnected methods that will throw exceptions if we don't have a principal unit name. The name of the principal unit has to be fetched from the juju-info relation (which is what makes the subordinate inject into the principal unit). However, the field in the relation is only populated on the relation_joined event so we can not use it in the setup phase. There are event handlers that use the principal unit name that are run in the setup phase, so it is guaranteed to fail if we don't check for the presence of the principal unit. I will write something more condensed.